Add reusable modal component.

aaronfc opened the pull request:
Fixes: 1446-gh-Automattic/sensei-pro

### Changes proposed in this Pull Request

* Implementation based on the same concept of sensei-modal used in Contact Teacher.


### Testing instructions
* See steps and video here: 1552-gh-Automattic/sensei-pro

### Discussion
- Should we try to replace the current Contact Teacher implementation with this new module? This is a component so it might need some re-working of how Contact Teacher works right now.
- I am using the `@wordpress/icon` components – it is working (as you can see in the linked pull request video) but maybe we should avoid that? @renatho ?

aaronfc committed:
Add Modal reusable component.

- Implementation based on the same concept of sensei-modal used in
  Contact Teacher.

-

aaronfc committed:
Set customisable title for Modal.

aaronfc committed:
Set fixed height for modal title.

alexsanford reviewed (commented):
Was there any cross-talk between the development of this PR and #5361? Maybe there is a need for completely separate components, but it seems like there is some amount of overlap.

I suppose maybe not, since #5361 is a confirmation dialog meant for the editor and this is a modal meant to be shared between the editor and the frontend. But I think it is at least worth a look to make sure we're not duplicating efforts.

fjorgemota commented:
Hey there.

No, there was no cross-talk between the development of this PR and #5361, mainly because #5361 just uses a modal instead of reimplementing it.

With this said, I don't see any overlap between this PR and #5361. I see an opportunity for improvement for #5361, though, where we could use this modal instead of the one provided by Gutenberg. However, I think it's only worth doing this if the confirmation dialog was going to be also shown on the frontend, which I don't think it's the case.

renatho commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,44 @@
+ */
+import classNames from 'classnames';
+
+const Modal = ( { isOpen, setIsOpen, title, children } ) => {
```
It would be nice to add some JSDoc here.

aaronfc replied:
Done! 130830568

renatho commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,44 @@
+
+const Modal = ( { isOpen, setIsOpen, title, children } ) => {
+	const close = ( event ) => {
+		setIsOpen( false );
```
WDYT about using a close callback only? I think it would make more sense since we'll never open the modal through the modal.

aaronfc replied:
Makes sense! Done in 130830568

renatho commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,44 @@
+	return (
+		<div
+			className={ classNames( 'sensei-modal', {
+				'sensei-modal--open': isOpen,
```
We could probably start the component with a `if ( isOpen ) { return null; }`. WDYT? Since we're using react, I think it would be the best approach for this type of cases because we don't need to mount the component if we won't show that. Unless I'm missing something here?

aaronfc replied:
Yeah, that makes much more sense. I was carried on by the previous implementation.

Done in 130830568

renatho commented on assets/shared/components/modal/style.scss:
```diff
@@ -0,0 +1,64 @@
+.sensei-modal {
+
+	display: none; // Hide modal by default.
```
Related to the other comment of returning `null` when the modal is not open. We probably don't need it.

aaronfc replied:
Removed! 130830568

renatho commented on assets/shared/components/modal/style.scss:
```diff
@@ -0,0 +1,64 @@
+		cursor: pointer;
+	}
+
+	&.sensei-modal--open {
```
Also, related to the other comments about returning null, we also probably don't need it. And I think we could also remove it from the `className`, right?

aaronfc replied:
Agree :D Done in 130830568

renatho commented on assets/shared/components/modal/style.scss:
```diff
@@ -0,0 +1,64 @@
+		overflow: hidden;
+		z-index: 2000;
+		width: 700px;
+		max-width: 96%;
```
Maybe it could be fullscreen for mobile. WDYT?

aaronfc replied:
Yeah, maybe? Let's keep an eye on this thread p6rkRX-3VB-p2 and let's confirm with Andrei how it should look like. I am also a little concerned about the "actions" available in the modals in the latest designs.

renatho commented on assets/shared/components/modal/style.scss:
```diff
@@ -0,0 +1,64 @@
+		font-weight: 600;
+		line-height: 24px;
+		border-bottom: 1px solid #ddd;
+		height: 24px;
```
Do we really need the `height` here? I'm wondering about cases where this title would break the line.

aaronfc replied:
This was for the case when title is empty. I replaced it with a `min-height` instead of `height` here: 130830568

WDYT?

renatho reviewed (commented):
> I am using the @wordpress/icon components – it is working (as you can see in the linked pull request video) but maybe we should avoid that? @renatho ?

I think that's fine! I checked it, and it's not enqueued by WP, but added to the bundle, and it's super small.

When you need. to check it, you can just run the `npm run build:analyzer` and check the dependencies.

And about the WP enqueues, you can see the dist files, like this one for this case: `assets/dist/interactive-blocks/interactive-blocks-frontend.asset.php`. ([Check here if you want more details](https://www.npmjs.com/package/@wordpress/dependency-extraction-webpack-plugin)).

![Screen Shot 2022-07-21 at 12 02 20](https://user-images.githubusercontent.com/876340/180247265-77be3431-9411-4bee-a153-b5ba5f22a1ea.png)

aaronfc committed:
Review improvements.

- Add jsdoc.
- Simplify hidden modal by returning `null` instead of hiding with css class.
- Receive close callback because modal won't be ever open from the modal
  itself.
- Fix css for height.

aaronfc reviewed (commented):
Added some minor changes.

Still pending to decide what the design should look like for mobile and if we need to implement something special for "actions".

aaronfc committed:
Update with `trunk`.

aaronfc committed:
Pass click event on handleClose callback for modal and css.

fjorgemota commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,45 @@
+			<button
+				className="sensei-modal__overlay"
+				aria-label="Close"
+				onClick={ ( e ) => handleClose( e ) }
```
Do we need this closure here?

I feel like something like this is already enough:

```suggestion
				onClick={ handleClose }
```

aaronfc replied:
You are right, removed here 036e3e3ba

fjorgemota commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,45 @@
+					<div className="sensei-modal__title">{ title }</div>
+					<button
+						className="sensei-modal__close-button"
+						onClick={ ( e ) => handleClose( e ) }
```
Again, do we need this closure here?

I feel like something like this is already enough:

```suggestion
						onClick={ handleClose }
```

aaronfc replied:
Removed here 036e3e3ba

fjorgemota commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @param {Object}   props
+ * @param {boolean}  props.isOpen      Whether the modal should be open or not.
+ * @param {Function} props.handleClose Callback to run when trying to close the modal.
```
WDYT about renaming this prop to `onClose`? 

It makes more sense considering that it's only an event handler in the current state. :)

aaronfc replied:
Done! b489cdaa8

fjorgemota reviewed (commented):
Hey there, nice PR overall! 

I found an issue while testing this component with 1572-gh-Automattic/sensei-pro with the Twenty Twenty One theme that I think might be worth taking a look.

Basically, this is how the modal appears to me while on Twenty Twenty One (notice that the background also changes while hovering over the modal vs. hovering on the overlay, besides the weird close button):


https://user-images.githubusercontent.com/529864/180857650-909c73b8-1137-430e-b24b-12af34aeecd0.mp4

At first, I thought it was Firefox being weird, but the same issue also happens on Chrome.

It's worth mentioning that this issue doesn't occur on the editor, and neither occurs when using the theme Twenty Twenty:

https://user-images.githubusercontent.com/529864/180858168-27505a89-cd46-4eea-94f4-00d484221c03.mp4

The only thing a little weird about this modal on Twenty Twenty theme is that the close button looks...misaligned? Maybe it's worth checking, too, considering that on your video here 1552-gh-Automattic/sensei-pro that doesn't happen.

Apart from that, I left some small comments about the close callback.

Also, it's worth asking: Should this modal still be open if the user presses ESC? Usually, when the user press ESC, the Gutenberg modal calls `onClose` and so the modal is closed properly, so that's why I'm asking. :)

aaronfc committed:
Remove closures and fix some styles for Twenty Twenty One.

aaronfc committed:
Add support to modal close upon ESC key pressed.

aaronfc committed:
Renaming `handleClose` to `onClose` for modal.

aaronfc commented:
> The only thing a little weird about this modal on Twenty Twenty theme is that the close button looks...misaligned? Maybe it's worth checking, too, considering that on your video here 1552-gh-Automattic/sensei-pro that doesn't happen.

I tested with Twenty Twenty after the latest changes in CSS and I don't see the misalignment anymore :)
<img width="759" alt="image" src="https://user-images.githubusercontent.com/799065/180967725-7748365a-d865-4d47-a138-937248361fb8.png">

> Also, it's worth asking: Should this modal still be open if the user presses ESC? Usually, when the user press ESC, the Gutenberg modal calls `onClose` and so the modal is closed properly, so that's why I'm asking. :)

I added support for ESC here: ccb8fe130 :D thanks!


Let me know if you see any other strange thing @fjorgemota

aaronfc committed:
Fix height for modal title.

aaronfc committed:
Fix height finally.

fjorgemota commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,60 @@
+	useRunOnEscape( onClose );
+
+	return (
+		<div className={ 'sensei-modal' }>
```
I think here we can refer to the `className` without the `{` and `}`, right?

Like this:

```suggestion
		<div className="sensei-modal">
```

fjorgemota commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,60 @@
+const useRunOnEscape = ( onClose ) => {
+	useEffect( () => {
+		const handleEsc = ( event ) => {
+			if ( event.keyCode === 27 ) {
```
hmmm, WDYT about maybe putting `import { ESC } from "@wordpress/keycodes"` and using that constant, `ESC`, here?

It would also be nice to verify if `event.defaultPrevented` is false, too.  Something like: `event.keyCode === ESC && !event.defaultPrevented`. 

WDYT?

aaronfc replied:
Changed to use ESCAPE here: a327b8bba

About the check for "defaultPrevented" I am not completely sure... that's supposed to prevent the default behaviour, right? But our "hide the modal" is not the default behaviour when pressing any key. So I am not sure we should check for that. Any extra opinion on this? Maybe I am missing the meaning of "preventDefault".

fjorgemota replied:
About the defaultPrevented, I'm being inspired mostly by Gutenberg, which does the same thing here: https://github.com/WordPress/gutenberg/blob/1339b969bf6c700ca7ee315fbb0959bbe0767e12/packages/components/src/modal/index.js#L102-L104

I think the reasoning for using `e.defaultPrevented` vs. `e.stopPropagation()` is to still allow the event to bubble but skip its handling in the overlay if the "default behavior" is prevented. Naturally, in most situations, the default behavior here would be handled by the browser (which would do nothing, in this case), but with a child node, and in a situation where we have the event bubbling from a child node up to its parent nodes, we would recognize "preventing the default behavior" as "just skip closing the modal".

Am I making any sense? :thinking:

aaronfc replied:
I think you might be right. I finally added it when I saw it was already done in Gutenberg: 0c029ec98

fjorgemota commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,60 @@
+			}
+		};
+		// Attach close event on Escape key.
+		// eslint-disable-next-line @wordpress/no-global-event-listener
+		window.addEventListener( 'keydown', handleEsc );
```
I think if you add the event listener on `document.body`  you'll not need this `eslint-disable-next-line` comment. :thinking:

aaronfc replied:
I tested and you were right! But finally I implemented it in a different way: 0c029ec98

fjorgemota commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,60 @@
+
+		return () => {
+			// Detach from keydown on component unmounting.
+			// eslint-disable-next-line @wordpress/no-global-event-listener
+			window.removeEventListener( 'keydown', handleEsc );
```
Same thing as above: I think if you add the event listener on `document.body`, you'll not need to disable eslint check.

renatho replied:
I was checking [the rule](https://github.com/WordPress/gutenberg/pull/26810), and I think the expected approach would be adding a ref in an element, and then call the `ownerDocument` from that.

I also just checked that Gutenberg is using a hook called [`useRefEffect`](https://github.com/WordPress/gutenberg/tree/trunk/packages/compose#userefeffect) for this type of events. See [an example here](https://github.com/WordPress/gutenberg/blob/fd0daedb128399bcb99d2e1829187ed0e70ebd2a/packages/block-editor/src/components/copy-handler/index.js#L159).

In this case, I think it would be something like this:

```js
const ref = useRefEffect( ( node ) => {
  const handleEsc = 
  ...
  node.ownerDocument.defaultView.addEventListener( 'keydown', handleEsc );
  ...
}, [ onClose ] );

return <div className="sensei-modal" ref={ ref }>...</div>
```

renatho replied:
Or maybe we don't even need add the listener to the window. I was checking that Gutenberg modal attaches it to [the overlay](https://github.com/WordPress/gutenberg/blob/v13.7.3/packages/components/src/modal/index.js#L132).

aaronfc replied:
Reimplemented the modal using the same approach than Gutenberg's modal: 0c029ec98

aaronfc replied:
asdfasdf

fjorgemota reviewed (commented):
Nice work! Just a few additional comments. =)

renatho commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,60 @@
+		<div className={ 'sensei-modal' }>
+			<button
+				className="sensei-modal sensei-modal__overlay"
+				aria-label="Close"
```
We need to add the `__` here to make it translatable.

renatho replied:
Oops! Now that I noticed it's the overall. I think it's not very good from an accessibility perspective. Maybe we could try to use the same as Gutenberg modal? The `__experimentalUseFocusOutside`.

aaronfc replied:
Reimplemented modal using your suggestions here: 0c029ec98

renatho commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,60 @@
+ * @param {Function} props.onClose  Callback to run when trying to close the modal.
+ * @param {string}   props.title    The title for the modal. Empty by default.
+ * @param {Object}   props.children The content of the modal.
+ * @return {JSX.Element} The modal component or null (if not open).
```
I think we don't need the `return` here. At some point, we had decided to not add `@return` to the React components. But if think it's especially important here, feel free to keep it. 😉

aaronfc replied:
Removed in here: 1d4574af6

renatho commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,60 @@
+			<div className="sensei-modal__wrapper">
+				<div className="sensei-modal__header">
+					<div className="sensei-modal__title">{ title }</div>
+					<button
```
I think we can add the `aria-label` here to this button to make it accessible.

aaronfc replied:
Done in da7e3965d

renatho commented on assets/shared/components/modal/index.js:
```diff
@@ -0,0 +1,60 @@
+ * @param {Object}   props.children The content of the modal.
+ * @return {JSX.Element} The modal component or null (if not open).
+ */
+const Modal = ( { onClose, title = '', children } ) => {
```
Another thing that I realized while checking the Gutenberg Modal is that we're not using Portals here. I think it's important for modal implementations, in order to make it work in any place (if it's in a position relative, in a overflow hidden, or any other weird context).

See it here: https://github.com/WordPress/gutenberg/blob/v13.7.3/packages/components/src/modal/index.js#L124

aaronfc replied:
Added portals in here too: https://github.com/Automattic/sensei/commit/0c029ec980f35536f4fcad1481f15aa8773ba4f3

renatho reviewed (commented):
Thank you for the improvements, Aarón! I did a new round of review, and added some extra suggestions here. 😉

aaronfc committed:
Inline classname.

Co-authored-by: Fernando Jorge Mota <fjorgemota@users.noreply.github.com>

aaronfc committed:
Use ESCAPE constant.

aaronfc committed:
Reimplement Modal to track focus and process keypress events.

- Reimplemented following a similar approach to the Gutenberg original
  Modal component.
- Modal is focused when component is loaded so that keypress events are
  handled throw the `onKeyDown` attribute.
- Modal is closed when unfocused – replacing the overlay as a button
  with a normal div.

aaronfc committed:
Removing return not needed in jsdoc.

aaronfc committed:
Adding aria-label to close button.

renatho commented on assets/shared/components/modal/style.scss:
```diff
@@ -0,0 +1,69 @@
+
+	color: #1E1E1E;
+
+	& &__overlay {
```
Is the extra `& ` needed in all classes? Maybe to override some style?

aaronfc replied:
Not needed in this case specifically but I needed it in some styles like the button, so I decided to add them in all of them.

renatho reviewed (dismissed):
Looks good and works well! I just needed to tweak a little the original Sensei Pro branch to test. haha 🙂

I just left a final comment, but if has a purpose, feel free to just ignore it. 😉

aaronfc commented:
I was doing some testing and I found that the new implementation breaks when opening the Media Library modal when adding, for example, an Image.

It breaks because, when the Media library our modal is unfocused (and closed) making the content of the Media Library modal not appear (not completely sure why).

I am now wondering if maybe we should go back to the <button> overlay... what do you think, @renatho ?

In order to test the issue:
- Add an Image.
- Use the Media Library.
- Media Library will not load.

aaronfc committed:
Unsettingn outline for close button.

renatho reviewed (approved):
Just to document our recent conversation while approving this:

As we talked, I think we should ignore [this problem](https://github.com/Automattic/sensei/pull/5367#issuecomment-1198083699) since we shouldn't have more than one modal open at the same time.

And on the editor side, we'll probably use the Popover component to keep aligned with the new design and avoid this issue.
